AUTHORING_GUIDE.md

```diff
@@ -281,216 +191,203 @@ Args:
 ```
 
-Note the name of the function matches the URL route. The docstring is kept
-simple because reading the function body reveals how the parameters are
-used.
+If a sample has testing requirements that differ from its runtime requirements
+(such as dependencies on [pytest](http://pytest.org/en/latest/) or other
+testing libraries), the testing requirements may be listed in a separate
+`requirements-testing.txt` file instead of the main `requirements.txt` file.
 
-Use `print` or `logging.info` in request handlers to print useful information
-as needed.
+## Region Tags
 
-### Argparse section
+Sample code may be integrated into Google Cloud Documentation through the use
+of region tags, which are comments added to the source code to identify code
+blocks that correspond to specific topics covered in the documentation. For
+example, see
+[this sample](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/cloud-sql/mysql/sqlalchemy/main.py)
+— the region tags are the comments that begin with `[START` or `[END`.
 
-For command-line samples, you'll need an argparse section to handle
-parsing command-line arguments and executing the sample functions. This
-section lives within the `if __name__ == '__main__'` clause:
+The use of region tags is beyond the scope of this document, but if you’re
+using region tags they should start after the source code header
+(license/copyright information), imports, and global configuration such as
+initializing constants.
 
-```python
-if __name__ == '__main__':
-    parser = argparse.ArgumentParser(
-        description=__doc__,
-        formatter_class=argparse.RawDescriptionHelpFormatter)
-```
+## Exception Handling
 
-Note the use of `__doc__` (the module-level docstring) as the description for
-the parser. This helps you not repeat yourself and gives users useful
-information when invoking the program. We also use `RawDescriptionHelpFormatter`
-to prevent argparse from re-formatting the docstring.
+Sample code should use standard Python exception handling techniques as
+covered in the [Google Python Style
+Guide](http://google.github.io/styleguide/pyguide.html#24-exceptions).
 
-Command-line arguments should generally have a 1-to-1 match to function
-arguments. For example:
+# Testing Guidelines
 
-```python
-parser.add_argument('source_file_name')
-parser.add_argument('destination_blob_name')
-```
+Samples should include tests to verify that the sample runs correctly and
+generates the intended output. Follow these guidelines while writing your
+tests:
+
+* Use [pytest](https://docs.pytest.org/en/latest/)-style tests and plain
+asserts. Don't use `unittest`-style tests or `assertX` methods.
+* Whenever possible, tests should avoid taking a dependency on the precise
+output of APIs. For example, if a test is intended to verify a JSON payload
+returned from an endpoint, it should only check for the existence of the
+expected keys and values, and the test should continue to work correctly
```

Comment:
What's the difference between the precise output vs. expected keys and values? (I'm guessing it's order, but I want to double check)

Reply:
For ML APIs in particular, the results may not be consistent. In these situations we should test that the call was successfully, but not necessarily that the data returned was consistent.

Reply:
Clarified that dependency on order should be avoided, and added guidance on just verifying API call was successful in some cases. Addressed in commit [751c8e2](https://github.com/GoogleCloudPlatform/python-docs-samples/pull/2728/commits/751c8e27c591a489d2cb47477b0a971d41fcf36a)